hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java

```diff
@@ -796,18 +802,19 @@ private void startJoin(OnJoinOp preJoinOperation) {
 
                 OnJoinOp preJoinOp = preparePreJoinOps();
 
-                if (preJoinOperation != null) {
-                    nodeEngine.getOperationService().run(preJoinOperation);
-                }
-
                 // post join operations must be lock free, that means no locks at all:
                 // no partition locks, no key-based locks, no service level locks!
                 OnJoinOp postJoinOp = preparePostJoinOp();
 
                 // this is the current partition assignment state, not taking into account the
                 // currently joining members
                 PartitionRuntimeState partitionRuntimeState = partitionService.createPartitionState();
-                for (MemberInfo member : joiningMembers.values()) {
+                for (BiTuple<MemberInfo, OnJoinOp> tuple : joiningMembers.values()) {
+                    MemberInfo member = tuple.element1();
+                    // Run all joining members' pre join operations
+                    if (tuple.element2() != null) {
+                        nodeEngine.getOperationService().run(tuple.element2());
```

Comment:
shouldn't we run these on-join-ops before preparePreJoinOps() in above order? Otherwise some state may not be propagated to non-master nodes.

Reply:
Good catch @ahmetmircik, thank you - I've moved the `PreJoinOp` execution back to its original location, just now it executes all ops instead of the latest only 👍

Reply:
I meant moving it before this line --> `OnJoinOp preJoinOp = preparePreJoinOps();`

Because `preparePreJoinOps` propagates master's state. If we run ops after it, the `preJoinOp` may miss some state hence it can't propagate it. This needs investigation though, but on the negative side, this will also mean, propagating same state back to the joining member if on-join-op is not idempotent.

Reply:
The original code was executing the passed `OnJoinOp` at the location I moved the batch handling to - since that's how it operated previously (only difference being it only executed 1 op instead of multiple if they exist). My aim was to preserve existing functionality.

Reply:
Given that `OnJoinOp` execution does remote calls to all members. As far as i see here, with the changes in this PR, we are introducing quadratic complexity for on-join operations executions. Like if you have 10 nodes, you will create 1_00 remote calls during joins, if you have 1_00 nodes, you will end up 10_000 remote calls etc.

Ideal solution seems, first executing operations wrapped in an `OnJoinOp` on master node while joining, so no remote calls can happen in this case then calling `preparePreJoinOps` of master node as i wrote above. With this way we can get rid of this quadratic complexity.

WDYT?

Reply:
> As far as i see here, with the changes in this PR, we are introducing quadratic complexity for on-join operations executions

This PR restores functionality to its previous state prior to the December 2022 commit - you're correct that the operation has quadratic complexity, but this is not new, it was quadratic prior to December 2022 as well. Also, even in its current state, if all members join sequentially (allowing all of the `OnJoinOp`s to be executed on master due to no batching), we have this same quadratic complexity.

From my experimentation with your suggestion however, I do not see any tangible change to moving these operations around - the `OnJoinOp`s being executed locally (on master) are the `OnJoinOp`s from joining members, while the `preparePreJoinOps` prepared are to be executed on the remote members. Regardless of which is executed first, the number of `OnJoinOp` operations run across the cluster remains the same.; they do not seem to have any correlation to each other (from my understanding, perhaps there's a piece missing from my POV).

Reply:
I meant with this PR it became worse, before there was only 1 on-join-operation during batch execution, now we have on-join-ops from all over the batched members called remotely. 

To clarify what i meant, i prepared a coded version of it: 
```
for (BiTuple<MemberInfo, OnJoinOp> tuple : joiningMembers.values()) {
                    if (tuple.element2() != null) {
                        OnJoinOp onJoinOp = tuple.element2();
                        try {
                            onJoinOp.beforeRun();
                        } catch (Exception e) {
                            throw new RuntimeException(e);
                        }
                        Collection<Operation> operations = onJoinOp.getOperations();
                        for (Operation operation : operations) {
                            nodeEngine.getOperationService().run(operation); <-- this only runs locally.
                        }
                    }
                }

 OnJoinOp preJoinOp = preparePreJoinOps();
```

with this kind of change, we will not have O(n*n), and it will only be O(n). Does this make sense?

Reply:
Thanks @ahmetmircik, your example helped a lot, I was completely overlooking the re-broadcasting that takes place within the `OnJoinOp#run` method - now that I've scoped that out, I completely understand what you're saying, and you're completely correct of course. I've made changes in this commit: https://github.com/hazelcast/hazelcast/pull/25114/commits/bccf40268bf36f61d7422c43d3560eb72b809559

Debugging shows a significant reduction in the number of `OnJoinOp` executions now, with tests still passing and functionality unaffected👍 Definitely a great catch, thank you!

Reply:
good catch @ahmetmircik! 💯